File: java/leetcode_grind/Day673.java
package leetcode_grind;

public class Day673 {
    // https://leetcode.com/problems/shortest-palindrome/description/?envType=daily-question&envId=2024-09-20
    static class Solution1 {
        public String shortestPalindrome(String s) {
            if (s.isEmpty())
                return s;
            var isPalindrome = new Object() {
                boolean apply(int i) {
                    int j = 0;
                    int end = i / 2;
                    while (i >= end) {
                        if (s.charAt(i) != s.charAt(j)) {
                            return false;
                        }
                        i--;
                        j++;
                    }
                    return true;
                }
            };
            int n = s.length();
            int max = 0;
            for (int i = 0; i < n; i++) {
                if (isPalindrome.apply(i)) {
                    max = i;
                }
            }
            String rest = s.substring(max + 1);
            var r = new StringBuilder(rest).reverse().toString();
            return r + s;
        }
    }

}

File: src/leetcode_grind/day_673.rs
// https://leetcode.com/problems/shortest-palindrome/description/?envType=daily-question&envId=2024-09-20
pub fn shortest_palindrome(s: String) -> String {
    fn build_prefix_table(s: String) -> Vec<usize> {
        let s = s.chars().collect::<Vec<_>>();
        let mut prefix_table = vec![0; s.len()];
        let mut length = 0;

        for i in 1..s.len() {
            while length > 0 && s[i] != s[length] {
                length = prefix_table[length - 1];
            }
            if s[i] == s[length] {
                length += 1;
            }
            prefix_table[i] = length;
        }

        prefix_table
    }

    let rev_s = s.chars().rev().collect::<String>();
    let comb_s = s.clone() + "#" + &rev_s;
    let mut prefix_table = build_prefix_table(comb_s.clone());
    let palindrome_length = prefix_table[comb_s.len() - 1];
    let suffix = s[palindrome_length..].chars().rev().collect::<String>();
    suffix + &s
}
